Added Waterfall terrain overlay animations

Mailaender opened the pull request:
Like tunnel tops it works best if terrain overlays https://github.com/OpenRA/OpenRA/issues/13929 are added as actors. The downside is that they have to get placed manually. However map editor/import tooling can probably automatize this in the future.

The tricky part was to get the offset right. Luckily the first frame includes a reference piece from terrain artwork. I took a screenshot, moved it and counted the pixels. It then turned out that the OpenRA `sequence.yaml` offset currently matches `art.ini` in that
* `TileXOffset=` is the same as our `X`
* `TileYOffset=` minus 24 is our `Y`

I don't know if that applies generally.

I added them to maps where I could spot waterfalls with both snow and temperate test cases included. Waterfalls-B, -C and -D are still missing.

Mailaender committed:
Add Waterfall terrain overlay animations.

reaperrr commented:
> Waterfalls-B, -C and -D are still missing.

In my opinion it would be better if these were added with this PR, too, even if that means it takes longer to get completed and merged.

Mailaender commented:
I wanted to wait for feedback first. It is usually terrible to invest many hours only to get rejected.

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -1084,6 +1084,23 @@
+		UseClassicPerspectiveFudge: False
+		QuantizedFacings: 1
+	Building:
+		Dimensions: 1, 1
```
IMO this should be set to match the actual size of the waterfall - e.g. 3x1.

Mailaender replied:
That will make the correct placement impossible. Similar to the mess we currently have with tunnel tops in the map editor.

pchote replied:
I ended up having to redo the tunnel top definitions to give them a proper footprint, like this, to solve that mess.

Mailaender replied:
Actually the reference point is a 1x1 cell here. Compare the first frame of the waterfall, which is a small square from neighboring cliffs etc.

pchote replied:
The animations should freeze when under the fog, and `FrozenUnderFog` requires a proper footprint to be set up.

pchote replied:
Actually `FrozenUnderFog` is overkill - `HiddenUnderFog` would have less performance overhead and do the job just as well.  In any case, that also requires a proper footprint.

Mailaender replied:
But how do you solve the map editor placement. It will always glitch due to heights and I assume the footprint which is too large. Try importing "They All Float" and place the tunnel top definitions. Some, actually most, will be impossible to get right as you can't position them at the right cell.

pchote replied:
By fixing the actor placement tool in the editor to handle the edge case where the TopLeft cell is at a different height to the cell underneath the cursor.

reaperrr commented:
> I wanted to wait for feedback first. It is usually terrible to invest many hours only to get rejected.

Ok, fair enough. This can wait until we've ironed out any issues with the-A waterfalls (at first glance they look good, though).

pchote commented:
Right now the map editor doesn't even support placing waterfall *tiles* (because of the lack of height or slope adjustment), and so #13985 is not an issue in practice.  Our importer will place them directly.  Blocking this over #13985 guarantees this won't be merged, unless you're currently working on a fix for that yourself.

pchote commented:
My fix for #14000 had an unexpected side effect of fixing #13985, so a compromise is no longer necessary.  See #14006.

Mailaender committed:
Setup footprints and hide waterfall animations under fog.

Mailaender commented:
Rebased and adjusted the offsets with my pixel counting and pattern matching skills:

![image](https://user-images.githubusercontent.com/756669/30251500-8022b93c-9660-11e7-96db-44131cbac23e.png)

using the first reference frame:
```yaml
wa03x:
	idle:
		ZRamp: 1
		UseTilesetExtension: true
		Length: 1
		Start: 0
```

abcdefg30 reviewed (commented):
Looks like we could remove a lot of duplication in the sequence yamls by defining and inheriting a `^Waterfall` template. For example:
```
^Waterfall:
	idle:
		ZRamp: 1
		UseTilesetExtension: true
		Length: 7
		Start: 1
		Tick: 200
```

Mailaender committed:
Add a ^waterfall template for inheritance.

Mailaender commented:
Good idea. Done.

abcdefg30 commented on mods/ts/sequences/misc.yaml:
```diff
@@ -637,6 +637,33 @@ tuntop04:
 		Offset: -24, -49, 24
 
+^waterfall:
+		ZRamp: 1
```
You have one level of indentation too much here, as it looks.

> OpenRA.Utility(1,1): Error: Failed with exception: OpenRA.YamlException: Bad indent in miniyaml at ts|sequences/misc.yaml:641

Mailaender replied:
Sorry, that was sloppy. Fixed.

pchote commented:
The implementation now looks good to me, and the temperate waterfalls look great ingame.  The snow waterfalls in Karasjok look weird, but I don't see how that could be anything but an artwork issue.

Can you please now set up the other waterfall types?  Sunstroke also has waterfalls.

Mailaender committed:
Add waterfalls to Sunstroke

Mailaender commented:
Added Sunstroke waterfalls. Having a bit of a struggle getting the backwards waterfalls (C and D) to render after the actor is getting placed. The cliff rendering seems to swallow them.

pchote commented:
You will need to give them a z offset so that they are drawn (only just) above the terrain.

pchote commented:
@Mailaender any progress here?

Mailaender commented:
No, I am stuck. Don't know how to calculate the Z offsets. Trial and error also doesn't work.

pchote commented:
I will sort out the z offsets for you if you can work on the importer placement.

pchote commented:
...and also push the incomplete rules/sequences that you have already!

Mailaender commented:
I would like to give up here. Either merge this with just the working Waterfall A variants or close it so someone else can continue.

pchote commented:
Moving to the future milestone.
